google/cloud/storage/internal/service_account_credentials_test.cc

```diff
@@ -82,60 +84,49 @@ struct FakeClock : public std::chrono::system_clock {
-        return std::unique_ptr<char[]>(copy);
-      }));
-  // Add this matcher AFTER the AnyMatcher above; gmock matches in reverse
-  // order of how you defined matchers, and we want this specific string to be
-  // matched before being checked and matched against the AnyMatcher above.
-  EXPECT_CALL(*handle, MakeEscapedString(StrEq(GRANT_PARAM_UNESCAPED)))
-      .WillRepeatedly(Invoke([](std::string const& x) {
-        const std::string actual_grant(GRANT_PARAM_ESCAPED);
-        auto const escaped_str_size = actual_grant.size();
-        auto copy = new char[escaped_str_size + 1];
-        std::memcpy(copy, actual_grant.data(), actual_grant.size());
-        copy[escaped_str_size] = '\0';
-        return std::unique_ptr<char[]>(copy);
-      }));
-}
-
 /// @test Verify that we can create service account credentials from a keyfile.
 TEST_F(ServiceAccountCredentialsTest,
        RefreshingSendsCorrectRequestBodyAndParsesResponse) {
-  auto mock_http_request_handle =
-      MockHttpRequest::Handle(storage::internal::GoogleOAuthRefreshEndpoint());
-
-  SetExpectCallsForMakeEscapedString(mock_http_request_handle);
-
-  EXPECT_CALL(*mock_http_request_handle,
-              PrepareRequest(An<std::string const&>(), false))
-      .WillOnce(Invoke([](std::string const& payload, bool) {
-        EXPECT_THAT(payload, HasSubstr(EXPECTED_ASSERTION_PARAM));
-        // Hard-coded in this order in ServiceAccountCredentials class.
-        EXPECT_THAT(payload, HasSubstr(std::string("grant_type=") +
-                                       GRANT_PARAM_ESCAPED));
-      }));
-
-  EXPECT_CALL(
-      *mock_http_request_handle,
-      AddHeader(StrEq("Content-Type: application/x-www-form-urlencoded")));
-
+  auto mock_request = std::make_shared<MockHttpRequest::Impl>();
   std::string response = R"""({
       "token_type": "Type",
       "access_token": "access-token-value",
       "expires_in": 1234
   })""";
-  EXPECT_CALL(*mock_http_request_handle, MakeRequest())
+  EXPECT_CALL(*mock_request, MakeRequest())
       .WillOnce(Return(storage::internal::HttpResponse{200, response, {}}));
 
-  ServiceAccountCredentials<MockHttpRequest, FakeClock> credentials(
+  auto mock_builder = MockHttpRequestBuilder::mock;
+  EXPECT_CALL(*mock_builder, BuildRequest(_))
+      .WillOnce(Invoke([mock_request](std::string payload) {
+        EXPECT_THAT(payload, HasSubstr(EXPECTED_ASSERTION_PARAM));
+        // Hard-coded in this order in ServiceAccountCredentials class.
+        EXPECT_THAT(payload, HasSubstr(std::string("grant_type=") +
+                                       GRANT_PARAM_ESCAPED));
+        MockHttpRequest result;
+        result.mock = mock_request;
+        return result;
+      }));
+
+  std::string expected_header =
+      "Content-Type: application/x-www-form-urlencoded";
+  EXPECT_CALL(*mock_builder, AddHeader(StrEq(expected_header)));
+  EXPECT_CALL(*mock_builder,
+              Constructor(StrEq("https://accounts.google.com/o/oauth2/token")))
```

Comment:
Why not use the constant here [1]? That way we don't break the test when we eventually change it to the new token endpoint.

[1] https://github.com/GoogleCloudPlatform/google-cloud-cpp/blob/962600306999448ee94a0f55788ce3951a22f36c/google/cloud/storage/internal/credential_constants.h#L36

Reply:
Done.